Make invalid states unrepresentable, fix bug

ruuda opened the pull request:
We have a bug, reproduced by #58, where we reset the integration status of a pull request, but not the build status, causing an inconsistent state. We do not care about the build status of pull requests that are not integrated, we can make invalid states unrepresentable by moving the build status into the `Integrated` integration status. This fixes the bug, and also simplifies a few other pattern matches.

Unfortunately, because these types are heavily used in Hoff, especially in the test suite, the diff is rather large. It can be reviewed commit by commit to make it more approchable.

A next step to make more invalid states unrepresentable would be to also fold the approval status and integration status into one, such that only approved pull requests can be integrated. That would be a similar change as this one, but as the diff is already  large, I prefer to do that separately.

I did not (yet) try to preserve backwards compatibility in the schema. It could be done, but we can also just make a breaking change, Hoff will scan for new pull requests at startup. It would lose two things:

* Approval status
* Previous integration result sha’s, which are used to ignore push events that we caused ourselves

If we deploy it at a quiet moment, this should not be a problem, I think that’s easier than writing a backwards-compatible parser.

fatho commented on src/Project.hs:
```diff
@@ -253,11 +238,10 @@ classifyPullRequest pr = case approvedBy pr of
   Just _  -> case integrationStatus pr of
```
If a pull request can only be integrated if it was approved, is this another thing we would want to express in the ADTs?

ruuda replied:
Yes, I would like to do this, but not as part of this PR, as it would be a change of a similar size, again touching all the same places.

fatho commented on src/Logic.hs:
```diff
@@ -397,15 +396,22 @@ proceed state = case Pr.getIntegrationCandidate state of
-    BuildFailed     -> do
-      -- If the build failed, this is no longer a candidate.
-      leaveComment pullRequestId "The build failed."
+  case Pr.integrationStatus pullRequest of
```
The function `proceed` has this comment at the call site for `proceedCandidate`

> ```haskell
>  -- If there is a candidate, nothing needs to be done. TODO: not even if the
>  -- build has finished for the candidate? Or if it has not even been started?
>  -- Do I handle that here or in the build status changed event? I think the
>  -- answer is "do as much as possible here" because the events are ephemeral,
>  -- but the state can be persisted to disk, so the process can continue after a
>  -- restart.
> ```

What this function does doesn't look like "nothing", so I suppose that comment needs an update.

ruuda replied:
Good find. I removed the entire comment, the approach of “do as much as possible” in the to do comment is already the approach taken by `proceedCandidate`.

fatho commented on src/Logic.hs:
```diff
@@ -459,12 +464,14 @@ pushCandidate (pullRequestId, pullRequest) state = do
-      newHead   = assert (approved && succeeded) $ case status of
-        Integrated sha -> sha
-        _              -> error "inconsistent state: build succeeded for non-integrated pull request"
+      newHead   = assert approved $ case status of
+        Integrated sha BuildSucceeded -> sha
+        Integrated _ _notSucceeded ->
+          error "Trying to push a candidate for which the build did not pass."
+        _notIntegrated ->
+          error "Trying to push a candidate that is not integrated."
```
A big chunk of this function is now validating a condition that we know is true at the call site.
Maybe we can lift this knowledge into the type level?

ruuda replied:
The `assert approved` will go away later once we fuse the integration status and approval status.

For the `newHead` match, I am not sure. We could accept the sha as argument and let the caller do the match, because the caller is doing the match anyway, but then there is nothing that guarantees that the sha you pass in agrees with the actual state of the pull request. (There is a similar problem with the pull request id, either we pass only that and then the function assumes that a lookup will not fail, or we pass both the id and the pull request like now, but then there is no guarantee that the two agree.)

Refinement types in Liquid Haskell could solve this, but it doesn’t seem to be as easy as turning on a language pragma.

Maybe it could be done with GADTs, with a type tag for each of the integration statuses :thinking:

So it looks like the trade-off is between trusting the caller to pass in the result of the right match, or trusting the caller to only call when the preconditions are satisfied. I have a slight preference for the latter, because for violations we can call `error`, and that is easy to debug, whereas passing in the wrong arguments could lead to silent corruption.

What do you think?

fatho replied:
You're right, for debugging, those checks might still be useuful.
As long as these errors are properly handled, that's fine. That's what I hinted at with my top-level comment:

> Since there are still runtime errors, would it make sense to also already address the error handling in the main function to prevent hoff from locking up again should we ever encounter those?

---

Just exploring some other ideas here:

> For the newHead match, I am not sure. We could accept the sha as argument and let the caller do the match, because the caller is doing the match anyway, but then there is nothing that guarantees that the sha you pass in agrees with the actual state of the pull request.

The only way to get a Sha at that point is extracting it from the integration status
A newtype wrapper could act as an extra safety net there making wrong code look obviously wrong:

```haskell
newtype IntegratedCommit = IntegratedCommit Sha

data IntegrationStatus = ... | Integrated IntegratedCommit BuildStatus | ...

-- Then this looks correct and typechecks
    Integrated sha buildStatus -> case buildStatus of
      BuildPending   -> pure state
      BuildSucceeded -> pushCandidate (pullRequestId, pullRequest) sha state
      BuildFailed    -> ...

-- But this looks quite wrong
    Conflicted -> pushCandidate (pullRequestId, pullRequest) (IntegratedCommit someRandomSha) state

-- And this one doesn't type check
    Conflicted -> pushCandidate (pullRequestId, pullRequest) someRandomSha state

-- And this also looks wrong if you know what the business logic is supposed to be (as in: why would one push if the build has failed:)
    Integrated sha buildStatus -> case buildStatus of
      BuildFailed  -> pushCandidate (pullRequestId, pullRequest) sha state
```

For the latter to look even more wrong, `pushCandidate` which sounds quite vague (where is it pushed to?) could be renamed to `pushCandidateToMaster`.

And lastly, `pushCandidate` is not even exposed from the module.
At some point, you'll have to draw the boundary of where you verify invariants and where you trust invariants to be upheld (although assertions checking that those assumption are correct are a good idea for debuggability).


> but then there is nothing that guarantees that the sha you pass in agrees with the actual state of the pull request

Also, this might be something where [Ghosts of Departed Proofs](https://kataskeue.com/gdp.pdf) might come in handy again.

ruuda replied:
Okay, let’s keep it like this for now then, and re-evaluate once we fold the approval status and integration status. Maybe at that point it becomes feasible to inline this, and then the problem will go away entirely.

ruuda commented:
> Since there are still runtime errors, would it make sense to also already address the error handling in the main function to prevent Hoff from locking up again should we ever encounter those?

Yes, good point. I opened #60 for this.

ruuda commented:
I added another drive-by fix in e0e8994d4f86a7edd9059835a0279a34faffc9af and removed an unused function in 6ceda8e129b539b0955ca225110220411508f13c.

Changing target branch to master so merging will include #58.

ruuda commented:
@OpsBotPrime merge

OpsBotPrime commented:
Pull request approved by @ruuda, rebasing now.

OpsBotPrime commented:
Rebased as af94fa53a0fd3105996c0ac2d23ccb98097c43a1, waiting for CI …
